cli_tools/common/utils/logging/firelog_logger.go

```diff
@@ -0,0 +1,311 @@
+		o.FailureMessageWithoutPrivacyInfo = getAnonymizedFailureReason(err)
+	}
+
+	return &o
+}
+
+func getInt64Value(s string) int64 {
+	i, err := strconv.ParseInt(s, 0, 64)
+	if err != nil {
+		return 0
+	}
+	return i
+}
+
+func (l *FirelogLogger) runWithServerLogging(function func() (*daisy.Workflow, error)) *ComputeImageToolsLogExtension {
+	var event *ComputeImageToolsLogExtension
+
+	// Send log asynchronously. No need to interrupt the main flow when failed to send log, just
+	// keep moving.
+	var wg sync.WaitGroup
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		l.logStart()
+	}()
+
+	if w, err := function(); err != nil {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			event, _ = l.logFailure(err, w)
+			log.Println(event.OutputInfo.FailureMessage)
+		}()
+	} else {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			event, _ = l.logSuccess(w)
+		}()
+	}
+
+	wg.Wait()
+	return event
+}
+
+// RunWithServerLogging runs the function with server logging
+func RunWithServerLogging(action string, params InputParams, function func() (*daisy.Workflow, error)) {
+	l := NewFirelogLogger(action, params)
+	l.runWithServerLogging(function)
+}
+
+func (l *FirelogLogger) sendLogToServer(logEvent *ComputeImageToolsLogExtension) logResult {
+	return l.sendLogToServerWithRetry(logEvent, 3)
+}
+
+func (l *FirelogLogger) sendLogToServerWithRetry(logEvent *ComputeImageToolsLogExtension, maxRetry int) logResult {
+	logMutex.Lock()
+	defer logMutex.Unlock()
+
+	for i := 0; i < maxRetry; i++ {
+		// Before sending a new request, wait for a while if server asked to do so
+		if nextRequestWaitMillis > 0 {
+			nextRequestWaitMillis = 0
+			time.Sleep(time.Duration(nextRequestWaitMillis) * time.Millisecond)
+		}
+
+		logRequestJSON, err := constructLogRequest(logEvent)
+		fmt.Println(string(logRequestJSON))
+		if err != nil {
+			fmt.Println("Failed to log to server: failed to prepare json log data.")
+			return failedOnCreateRequestJSON
+		}
+
+		if !serverLogEnabled {
+			return serverLogDisabled
+		}
+
+		req, err := http.NewRequest("POST", l.ServerURL, bytes.NewBuffer(logRequestJSON))
+		if err != nil {
+			fmt.Println("Failed to create log request: ", err)
+			return failedOnCreateRequest
+		}
+
+		req.Header.Set("Content-type", "application/json")
+		req.Header.Set("X-Goog-Api-Key", reverse(key))
+		resp, err := httpClient.Do(req)
+		if err != nil {
+			fmt.Println("Failed to log to server: ", err)
+			continue
+		}
+
+		defer resp.Body.Close()
+		body, _ := ioutil.ReadAll(resp.Body)
+		var lr LogResponse
+		if err = json.Unmarshal(body, &lr); err != nil {
+			fmt.Println("Failed to parse log response: ", err, "\nResponse: ", string(body))
+			return failedToParseResponse
+		}
+
+		// Honor "NextRequestWaitMillis" from server for traffic control. However, wait no more than 5s to prevent a long
+		// stuck
+		if lr.NextRequestWaitMillis > 0 {
+			nextRequestWaitMillis = lr.NextRequestWaitMillis
+			if nextRequestWaitMillis > 5000 {
```

Comment:
By default, next_request_wait_millis is like 900000 (15 minutes), not 5 seconds.

Reply:
Oh that is tough. I saw it's 5s when I send a request to "FIRELOG_TEST".
The reason why our tool can't wait too long: it's not a long-live application. Instead, it's exit once it finished its task. Even worse, it's executed in cloudbuild as a docker, so the whole environment will be recycled immediately. We can't wait 15min to resend the log. That is why limit it to 5s. As a remedy, I limit the max retry to 3.
Do you have any suggestion on this?
What I can do is to limit the max retry to 2. Or, even 1 (which is no retry). (we are OK to lose some of the log, but at least try twice to avoid bad network situation from my perspective).